### PR TITLE
fix(page): add screen dpi property to page.pdf method

### LINKED ISSUE #666: PDF creation pixel sizes causes layout to change
Hi,

When creating PDFs from a responsive web page, lets say the page has a css rule to show a narrower version below `1024` pixels screen width and we supply `width: "1024px"` to `pdf` function. The rule should not apply on this case however it does and the layout changes in generated pdf. 

Only when the width parameter is **4/3** of the actual width of viewport, then the pdf has exact same layout of given viewport width. This seems to be something related to an inches / points confusion somewhere (96 pixels per inch, 72 pixels per point. 96/72 = 4/3)

So, when I provide `1365px` as pdf width I get the exact representation of the page at `1024px` viewport width. Anything below `1365px`, page acts like the viewport is below `1024px` and applies the appropriate responsive css rules.

I'm not sure if this is an upstream bug or something else though. Margins are set to 0 btw. 

Any ideas?

### LINKED ISSUE #3098: Pdf width is smaller than it should be
### Steps to reproduce
There seems to be an issue with the way the function page.pdf interprets the width of the page. Eventually using the design of the media query below the current one and I can't seem to know what's going on here.

**Tell us about your environment:**

* Puppeteer version: 1.5.0
* Platform / OS version: Mac
* URLs (if applicable): http://stephencaver.com/
* Node.js version: 8.11.3

**What steps will reproduce the problem?**
To reproduce the issue, you can try a page.pdf with this page http://stephencaver.com/, with a width of 902 and an height of 2000 (height does not matter here).
You'll see that the design display a two rows website demo instead of three, which is the rule for the media-query below 900 pixels.

_Please include code that reproduces the issue._
```
import puppeteer from 'puppeteer';
import fs from 'fs';
import express from 'express';
import bodyParser from 'body-parser';
import { join } from 'path';
import mkdirp from 'mkdirp';

const port = '8800';

/**
 * Creates and returns a configured express server.
 */

function createExpressServer() {
  let app = express();
  app.use(bodyParser.json({ limit: '50mb' }));
  app.use(bodyParser.urlencoded({ limit: '50mb', extended: true }));

  const wait = (timer) => new Promise(resolve => setTimeout(() => resolve(), timer));

  app.post('/createPDF/', async (req, res) => {
    const startTime = Date.now();
    console.log("Received request to render page: %s", req.body.url);

    // Launch Puppeteer
    const uuid = uuidv4();

    await page.emulateMedia('screen');

    // Go to the page and wait for all the connection on the page to be resolved
    await page.goto(`${req.body.url}`, {waitUntil: 'networkidle0'});

    // Send the response
    switch(req.body.format) {
      case 'pdf':
        await page.pdf({
          path: join(__dirname, PDF_EVIDENCE_FD, `${uuid}.pdf`),
          height: pageHeight,
          width: pageWidth,
          printBackground: true
        });

        res.status(200).sendFile(`${uuid}.pdf`);
        break;

      default:
        res.status(501).send({ error: 'error'});
        break;
    }
  });

  app.on("listening", function() {
    console.log("Listening on port %d", app.port);
  });
  return app;
}

let appInstance = createExpressServer();

appInstance.listen(port);
```

**What is the expected result?**
A three row demo section.

**What happens instead?**
A one row below section.

### PR DESCRIPTION
**Pull request by ggevgg**
This patch adds the dpi property to the page.pdf method to fix pdf output resolution

Fixes #3098, Fixes #666

### CODE REVIEW

File: docs/api.md [hunk 1]
```diff
@@ -1514,6 +1514,7 @@ Page is guaranteed to have a main frame which persists during navigations.
     - `bottom` <[string]|[number]> Bottom margin, accepts values labeled with units.
     - `left` <[string]|[number]> Left margin, accepts values labeled with units.
   - `preferCSSPageSize` <[boolean]> Give any CSS `@page` size declared in the page priority over what is declared in `width` and `height` or `format` options. Defaults to `false`, which will scale the content to fit the paper size.
+  - `dpi` <[number]> Screen dpi. Defaults to 96.
```
Comment (vsemozhetbyt) [inline docs/api.md]: Maybe:
```suggestion
  - `dpi` <[number]> Screen DPI. Defaults to 96.
```

**Comment by vsemozhetbyt** [comment]
Could these scary messages on approved suggestions be somehow addressed by the bot creators?

**Comment by JoelEinbinder** [comment]
Code looks reasonable, but I am not that familiar with the topic. How do I know what dpi to pass into this function? If it is supposed to match something in the website/chrome, we should be querying for it instead.

**Comment by ggevgg** [comment]
> Code looks reasonable, but I am not that familiar with the topic. How do I know what dpi to pass into this function? If it is supposed to match something in the website/chrome, we should be querying for it instead.

In most cases, Chrome correctly detects dpi, and you do not need to change the default dpi. But when the PDF resolution is not as expected, we need the ability to change this property.

**Comment by JoelEinbinder** [comment]
It looks like I can get the dpi for a web page via a script like this:
```js
const dpi = await page.evaluate(() => {
  const div = document.createElement('div');
  div.style.position = 'absolute';
  div.style.top = 0;
  div.style.left = 0;
  div.style.width = '1in';
  div.style.height = '1in';
  document.body.appendChild(div);
  const height = div.offsetHeight;
  div.remove();
  return height;
});
```

Should the DPI value that the user passes into this method always match the result of this function?

**Comment by ggevgg** [comment]
This approach has some minuses:
1. it isn't quite clean - it affects original DOM
2. our new DIV element can be affected by styles of original DOM and we can receive incorrect result

We have to provide a manual correction of the coefficient for the application flexibility. In some cases we need to use particular dpi and it may be different from the system one
Also we have checked this approach in our container. Result is the same as default value (equals 96), but isn't fit to us (manual dpi equals 72)

**Comment by kblok** [comment]
Just my two cents here.
I'm trying to answer [this question on StackOverflow](https://stackoverflow.com/questions/54784768/how-to-generate-pdf-of-exactly-specified-part-of-the-page-with-puppeter). I pulled this branch and used @JoelEinbinder code to determine my DPI. It was 96.

It seems that this solution isn't pixel perfect.
image

**Comment by clickyclick** [comment]
This wont fix the two bugs linked since the bugs occur even if you pass in inches.

**Comment by aslushnikov** [comment]
@ggevgg Spring cleanup! Sorry for this hanging for so long.

I don't think this code fixes the related bugs. Most of the pixel issues we have in PDF printing are related to the inherent unit translation on the chromium side; the fix should be on the chromium side as well.

> In most cases, Chrome correctly detects dpi, and you do not need to change the default dpi. But when the PDF resolution is not as expected, we need the ability to change this property.

I don't actually understand this. What are the cases?

**Comment by cliedelt** [comment]
> I don't think this code fixes the related bugs. Most of the pixel issues we have in PDF printing are related to the inherent unit translation on the chromium side; the fix should be on the chromium side as well.

are you able to fix this? @ggevgg @aslushnikov  its one of the highest rated bugs 😞 #2278

**Comment by aslushnikov** [comment]
> are you able to fix this?

@orgertot not yet.

Closing PR since I don't think it addresses the issue.

**Pull request closed without merging**